packages/shared/src/components/history/ReadingHistoryList.tsx

```diff
@@ -0,0 +1,75 @@
+import React, { ReactElement } from 'react';
+import classNames from 'classnames';
+import {
+  HideReadHistory,
+  ReadHistoryInfiniteData,
+} from '../../hooks/useReadingHistory';
+import classed from '../../lib/classed';
+import {
+  isDateOnlyEqual,
+  getReadHistoryDateFormat,
+} from '../../lib/dateFormat';
+import ReadingHistoryItem from './ReadingHistoryItem';
+
+const DateTitle = classed('h2', 'typo-body text-theme-label-tertiary');
+
+const getDateGroup = (date: Date, isFirstLabel: boolean) => {
+  const label = getReadHistoryDateFormat(date);
+
+  return (
+    <DateTitle
+      key={label}
+      className={classNames('px-6 my-3', isFirstLabel && 'mt-0')}
+    >
+      {label}
+    </DateTitle>
+  );
+};
+
+export interface ReadHistoryListProps {
+  data: ReadHistoryInfiniteData;
+  onHide: HideReadHistory;
+  infiniteScrollRef: (node?: Element | null) => void;
+}
+
+function ReadHistoryList({
+  data,
+  onHide,
+  infiniteScrollRef,
+}: ReadHistoryListProps): ReactElement {
+  let currentDate: Date;
+
+  return (
+    <div className="flex relative flex-col">
+      {data?.pages.map((page, pageIndex) =>
+        page.readHistory.edges.reduce((dom, { node: history }, edgeIndex) => {
```

Comment:
Maybe it's just me, but had to read this function twice.
Could we not either do all these calculations on the backend?

Or make this a separate function compared to doing it in the rendering?

Reply:
Like use a reduce function `mapReadHistoryToDays` or something

Reply:
Unfortunately, this is the necessary shape of the actual response from the backend since this is a paginated result.

We used to use `reduce` in the `UpvotedPopupModal` for this scenario but it was deemed redundant as an additional computation for this loop.

Reply:
But can't we extract the function from the render?
Just find it weird a render does calculations.

Reply:
Yes, I realized that as well. I have now applied the extraction of the actual function.